### ISSUE TITLE
Image upload hook action to kcfinder.

### ISSUE THREAD

**Issue by concept-core** [GetmeUK/ContentTools#199]
Hello,

I am looking for a way so I can upload a image using kcfinder from the contenttools script.
Since I am no way a hero with javascript I hope somebody could show me how to do it or push me in the right direction.

Best regards,
Wouter.

**Comment by anthonyjb** [comment]
Hi @concept-core, I'm afraid I'm not familiar with kcfinder at all but with a bit of luck someone here has and can be able to help. You might have more luck posing this question on StackOverflow though.

**Comment by concept-core** [comment]
Hi Anthony. 

Let me rephrase since you are better with js then me.
Is there a way I can double click the img icon or change the whole body of the image *modal* ? 

If I can double click the image I should be able to pass trough the url of the image I want to select.

And if I can change the contents of the *modal* I should be able to place a iframe of some sort there. 

If you want you can take a look at kcfinder integration.
That's done in js, maybe it can help you come up with a solution. 

Best regards, 
Wouter.

**Comment by anthonyjb** [comment]
Hi Wouter,

If you want to detect a double-click against images one approach would be to do the following:

```JavaScript
ContentEdit.Root.get().bind('mount', function(elm) {
    // We're only interested in images
    if (elm.type() != 'Image') { return; }

    // Add an event listener for double clicks
    elm.domElement.addEventListener('dblclick', function(ev) {
        ...call kcfinder...
    });
});
```

Let me know if that's of any use?

**Comment by concept-core** [comment]
Heej Anthony, 

Could we be able to talk on Skype tomorrow? 
I still cannot figure it out and I think my explanation goes nowhere.

If you think that is ok may I send you a direct mail.

Best regards, 
Wouter.

**Comment by anthonyjb** [comment]
Sure I'll be around tomorrow afternoon and evening, my skypename is anthonyblackshaw, if you add me we can try and touch base tm.

**Comment by anthonyjb** [comment]
OK so here's an example of a base integration which I'm hoping you can test for me and that we can improve on once we have it working in it's simplest form:

```JavaScript
// So this little bundle of variables is required because I'm using CoffeeScript
// constructs and this code will potentially not have access to these. Longer 
// term we can look to bring this into the core library as an integration option
// and rewrite it in CoffeeScript but (despite appearances) this is probably
// easier for the moment as you're not used to CoffeeScript.
var __slice = [].slice,
__indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; },
__hasProp = {}.hasOwnProperty,
__extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
__bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

// Define out KCFinder image tool
var KCFImageTool = (function(_super) {
    __extends(KCFImageTool, _super);

    function KCFImageTool() {
      return KCFImageTool.__super__.constructor.apply(this, arguments);
    }

    // Register the tool with ContentTools (in this case we overwrite the 
    // default image tool).
    ContentTools.ToolShelf.stow(KCFImageTool, 'image');

    // Set the label and icon we'll use 
    KCFImageTool.label = 'Image';
    KCFImageTool.icon = 'image';

    KCFImageTool.canApply = function(element, selection) {    
        // So long as there's an image defined we can alwasy insert an image
        return true;
    };

    KCFImageTool.apply = function(element, selection, callback) {

        // Configure the KCFinder application
        window.KCFinder = {
            callBack: function(url) {
                // Once the user has selected an image insert it

                // Create the image element
                var image = new ContentEdit.Image({src: url});

                // Insert the image
                var insertAt = KCFImageTool._insertAt(element);
                insertAt[0].parent().attach(image, insertAt[1]);

                // Set the image as having focus
                image.focus();

                // Call the given tool callback
                return callback(true);

                window.KCFinder = null;
            }
        };
        
        // Open the KCFinder
        window.open(
            '/demo/kcfinder/browse.php?type=files&dir=files/public', 
            'kcfinder_textbox',
            'status=0, toolbar=0, location=0, menubar=0, directories=0, ' +
            'resizable=1, scrollbars=0, width=800, height=600'
        );

    };

    return KCFImageTool;

})(ContentTools.Tool);
```

**Comment by concept-core** [comment]
For other people who want to use KCFinder this is the code anthony came up with.

Here is the code we used to get it working correctly.

```javascript
// So this little bundle of variables is required because I'm using CoffeeScript
// constructs and this code will potentially not have access to these. Longer 
// term we can look to bring this into the core library as an integration option
// and rewrite it in CoffeeScript but (despite appearances) this is probably
// easier for the moment as you're not used to CoffeeScript.
var __slice = [].slice,
__indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; },
__hasProp = {}.hasOwnProperty,
__extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
__bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

// Define out KCFinder image tool
var KCFImageTool = (function(_super) {
    __extends(KCFImageTool, _super);

    function KCFImageTool() {
      return KCFImageTool.__super__.constructor.apply(this, arguments);
    }

    // Register the tool with ContentTools (in this case we overwrite the 
    // default image tool).
    ContentTools.ToolShelf.stow(KCFImageTool, 'image');

    // Set the label and icon we'll use 
    KCFImageTool.label = 'Image';
    KCFImageTool.icon = 'image';

    KCFImageTool.canApply = function(element, selection) {    
        // So long as there's an image defined we can alwasy insert an image
        return true;
    };

    KCFImageTool.apply = function(element, selection, callback) {

        // Configure the KCFinder application
        window.KCFinder = {
            callBack: function(url) {
                // Once the user has selected an image insert it

                // Create the image element
                var image = new ContentEdit.Image({src: url, width: 400, height: 400, 'data-ce-max-width': 800});

                // Insert the image
                var insertAt = KCFImageTool._insertAt(element);
                insertAt[0].parent().attach(image, insertAt[1]);

                // Set the image as having focus
                image.focus();

                // Call the given tool callback
                return callback(true);

                window.KCFinder = null;
            }
        };

        // Open the KCFinder
        window.open(
            '/path-to-kcfinder/browse.php?type=images&dir=images/public', 
            'kcfinder_textbox',
            'status=0, toolbar=0, location=0, menubar=0, directories=0, ' +
            'resizable=1, scrollbars=0, width=800, height=600'
        );

    };

    return KCFImageTool;

})(ContentTools.Tool);
```
